Use udisksctl and expect to unlock encrypted volume without needing supeuser privileges

CristianCantoro opened the pull request:
Coming from this PR [[1]], this pull request does the same for this project.

One caveat, I could not test it because I am already using @agherzan's [yubikey-full-disk-encryption](https://github.com/agherzan/yubikey-full-disk-encryption), but it should work (granted, it has needed some adaptations given the differences between `bash` and `sh`).

As said in the [original PR][1], this project adds a dependency on [`expect`](https://linux.die.net/man/1/expect), which is packaged in Debian, its derivatives, and in many other distros.

[1]:  https://github.com/agherzan/yubikey-full-disk-encryption/pull/17

Vincent43 commented:
It may need adding `expect` and `udisks2` dependencies to https://github.com/cornelinux/yubikey-luks/blob/master/debian/control#L10

Vincent43 commented:
@CristianCantoro I believe you can test it very easily because password generating scheme is quite identical. The only thing you need to do (assuming you use password as challenge option) is to set `CONCATENATE=1` and `HASH=1` in https://github.com/cornelinux/yubikey-luks/blob/master/ykluks.cfg and copy it to `/etc` (or build and install package). Generated LUKS passwords will be identical.

CristianCantoro commented:
@Vincent43 said:
> @CristianCantoro I believe you can test it very easily [...]

The point was that I did not want to change my Yubikey configuration  (since you can not set a challenge with this project and you have to use your password), enroll a new challenge on the key, and add a new passphrase to the encrypted volume. I worked around all that hard-coding my challenge in the script while I was testing.

That said, I had to make a lot of modification because `yubikey-luks-open` uses [`/bin/sh`][2] which on Ubuntu and Debian is symlinked to `/bin/dash`, a POSIX-compliant shell in which some of the expressions that I have used in [`ykfde-open`][1] don't work anymore. I have added detailed comments about what the script is doing to explain that.

It has been way more difficult than it should have been, but in the end it works and I have tested it on my machine (Ubuntu 16.04 LTS).

I have also added the dependencies, in the debian package. Note that I can not test the build of the package (I will report a couple of bugs for that).

[1]: https://github.com/agherzan/yubikey-full-disk-encryption/pull/17/commits/314657e0d180090de6288fa56d51ceb71a320155#diff-ca4ed95aa62e9b234a8b97da5d793c2bR76
[2]:  https://github.com/cornelinux/yubikey-luks/blob/master/yubikey-luks-open#L1

Vincent43 commented:
Sorry, I thought that you are using password as challenge mode. Anyway I successfully tested your changes.

Vincent43 commented:
> I spotted an outright error with the initial if [ "$(id -u)" -eq 0 ] (it was -ne, before);

@CristianCantoro what do you mean? It was excatly same code you changed [before](https://github.com/agherzan/yubikey-full-disk-encryption/commit/f7bd77c4474a6d655489c8536d2e94bf7926a6ab#diff-ca4ed95aa62e9b234a8b97da5d793c2bL6) and I believe it was valid before your approach.

CristianCantoro commented:
@Vincent43 said:
> @CristianCantoro what do you mean? It was excatly same code you changed before and I believe it was valid before your approach.

The condition `[ "$(id -u)" -ne 0 ]` is true if the user id, ovbtained via the command `id -u` is *not equal* (`-ne`) to `0`, which is the value of user id of the superuser (`root`).

So, what the original code does is "check the user's id, if it is  different from `0`, then  the user is not `root`", but the script needs to be run by root, so we print an error message and exit with an error code (`exit 1`, any non-zero exit value is considered an error status).

What I am doing in the new version is the following:
* if the user id is *equal to* `0` - `[ "$(id -u)" -eq 0 ]` - then the user is root and we can use the `cryptsetup` command directly (which needs superuser privileges)
* otherwise (`else`) the user is not `root` and the `udisksctl` command is used, which means that the encrypted drive is unlocked by the user without needing administrative privileges.

That explains the difference, I my first version of this pull request I was checking`[ "$(id -u)"  -ne 0 ]` for "the user is root", which is wrong.

cornelinux commented:
What does acutally happen, when the user is not root?

Vincent43 commented:
When user is not root `udisksctl` is used for opening container instead of `cryptsetup`. It's similar how most graphical environments are handling opening LUKS containers.

CristianCantoro commented:
I have rebased the commits to the last changes.
